sdk/go/docker/doc.go

```diff
@@ -1,10 +1,178 @@
+}
+
+// RunCommandThatMustSucceed is used to determine if the full command line should be reported
+// when an error happens.  In general reporting the full command line is fine.  But it should be set
+// to false if it might contain sensitive information (like a username/password)
+func RunCommandThatMustSucceed(cmd string, args []string, logResource pulumi.Resource,
+	reportFullCommandLine bool, env map[string]string) (string, error) {
+
+	res := <-runCommandThatCanFail(cmd, args, logResource, reportFullCommandLine, false, env)
+
+	if res.err != nil {
+		return "", errors.Wrapf(res.err, "%s failed with error",
+			getCommandLineMessage(cmd, args, reportFullCommandLine, env))
+	}
+
+	return res.stdout, nil
+}
+
+// Runs a CLI command in a child process, returning a promise for the process's exit. Both stdout
+// and stderr are redirected to process.stdout and process.stder by default.
+//
+// If the [stdin] argument is defined, it's contents are piped into stdin for the child process.
+//
+// [logResource] is used to specify the resource to associate command output with. Stderr messages
+// are always sent (since they may contain important information about something that's gone wrong).
+// Stdout messages will be logged ephemerally to this resource.  This lets the user know there is
+// progress, without having that dumped on them at the end.  If an error occurs though, the stdout
+// content will be printed.
+//
+// The promise returned by this function should never reach the rejected state.  Even if the
+// underlying spawned command has a problem, this will result in a resolved promise with the
+// [CommandResult.code] value set to a non-zero value.
+func runCommandThatCanFail(cmdName string, args []string, logResource pulumi.Resource, reportFullCommandLine bool,
+	reportErrorAsWarning bool, env map[string]string) <-chan commandResult {
+
+	logging.InitLogging(true, 7, false)
+
+	// Let the user ephemerally know the command we're going to execute.
+	logging.Infof("Executing " + getCommandLineMessage(cmdName, args, reportFullCommandLine, env))
+
+	// Generate a unique stream-ID that we'll associate all the docker output with. This will allow
+	// each spawned CLI command's output to associated with 'resource' and also streamed to the UI
+	// in pieces so that it can be displayed live.  The stream-ID is so that the UI knows these
+	// messages are all related and should be considered as one large message (just one that was
+	// sent over in chunks).
+	//
+	// We use rand here in case our package is loaded multiple times in memory (i.e. because
+	// different downstream dependencies depend on different versions of us).  By being random we
+	// effectively make it completely unlikely that any two cli outputs could map to the same stream
+	// id.
+	//
+	// Pick a reasonably distributed number between 0 and 2^30.  This will fit as an int32
+	// which the grpc layer needs.
+	streamID := rand.Int()
+
+	r := make(chan commandResult)
+	go func() {
```

Comment:
I actually don’t think we need any concurrency here. I think we can let this function be purely synchronous. That will also simplify things meaningfully. 

Looking at the .NET variant of this - all calls to this function are `await RunCmmandThatMustSucceed` so there doesn’t appear to be any dependence on concurrency - which also matches my intuition that all of the functionality here is really intended to be synchronous. 

If we *do* decide there is value in having some concurrency at the `buildAndPush` level, we can likely introduce a single goroutine there instead of having to manage it at this low level (one of the benefits of God’s concurrency model over async/await).